Improve CharacterBody script templates.

yosoyfreeman opened the pull request:
This PR improves the default CharacterBody template for both 2D and 3D.  It fixes the following issues:

**Input distortion:**

The line:

`var direction := (transform.basis * Vector3(input_dir.x, 0, input_dir.y)).normalized()`

Performs the basis multiplication against a potentially non orthonormal basis.  This is often the result of an object which doesn't have an scale of 1. However, because the normalization happens after the possibly wrong basis multiplication, the direction of the original input is now lost. This can be observed if you scale one of the axis to be bigger and move diagonally.

This has been fixed by removing the normalization (Which is not necessary because get_vector always reports at max a unit vector) and using an orthonormal basis instead. The scale no longer alters the input direction. This also allowed to use an smaller deadzone for the joystick, allowing the template to work with controllers too and illustrate how is done.

**Acceleration had multiple issues:**

It was accelerating instantly, but slowing down is done by moving the speed towards zero in an FPS dependent way, as there is no use of delta involved. Also, accelerating per axis is geometrically incorrect (Same issue as with non normalized vectors) and causes a very unnatural feeling. Also, the fps deceleration was the max speed.

```
if direction:
		velocity.x = direction.x * SPEED
		velocity.z = direction.z * SPEED
	else:
		velocity.x = move_toward(velocity.x, 0, SPEED)
		velocity.z = move_toward(velocity.z, 0, SPEED)
```

Now we isolate the horizontal component of our velocity and perform correct acceleration/deceleration

```
# Move the current horizontal velocity towards the target horizontal velocity by the acceleration factor multiplied by delta.
horizontal_velocity = horizontal_velocity.move_toward(target_horizontal_velocity, ACCEL * delta)
```

I have also taken into account the up vector of the CharacterBody node. We were ignoring it and assuming always an up vector of ` 0 1 0`. Now everything is done having into account that the up vector is arbitrary.

I have tested it and works nicely. The c# versions are my first attempt with MONO, so i  would appreciate some help in those to make sure i followed conventions properly.

yosoyfreeman committed:
Improved CharacterBody script templates.

AThousandShips commented:
See also:
* https://github.com/godotengine/godot/pull/73873

AThousandShips commented on modules/gdscript/editor/script_templates/CharacterBody2D/basic_movement.gd:
```diff
@@ -4,24 +4,40 @@ extends _BASE_
 
 const SPEED = 300.0
+const ACCEL = 300.0
+
```
```suggestion
```
For consistency

AThousandShips commented on modules/gdscript/editor/script_templates/CharacterBody2D/basic_movement.gd:
```diff
@@ -4,24 +4,39 @@ extends _BASE_
+	# Get the input axis. As good practice, you should replace UI actions with custom gameplay actions.
+	var input_axis: float = Input.get_axis("ui_left", "ui_right")
+
+	# Calculate the intended direction in 3D space.
```
```suggestion
	# Calculate the intended direction in 2D space.
```
Should probably be "plane" or similar as there's no 2D space technically

AThousandShips commented on modules/mono/editor/script_templates/CharacterBody2D/basic_movement.cs:
```diff
@@ -3,40 +3,50 @@
+		// Get the input axis. As good practice, you should replace UI actions with custom gameplay actions.
+		float inputAxis = Input.GetAxis("ui_left", "ui_right");
+
+		// Calculate the intended direction in 3D space.
```
```suggestion
		// Calculate the intended direction in 2D space.
```
Same observation

yosoyfreeman commented:
Hi! I seen the proposal, but it contains some gameplay specific choices like walk/run, modifying the velocity when changing direction for game feel purposes etc.  As it's clear on the discussion, such things can be really specific and are hard to implement in a generic way. This proposal is meant to keep the same functionality we already have, just solving the input, acceleration and up direction issues. I also tried to make more clear how to properly handle acceleration along the horizontal plane and gravity-jump along the up direction.

So this proposal is more about fixing what we have and polishing it than it is about offering more features.

AThousandShips commented:
That's a PR with code changes to the same templates so good to get the context from those discussions

yosoyfreeman commented:
Thank you a lot for the review! I'll try to have it fixing as soon as possible!

PS: Yes i think the linked PR is relevant! I was just trying to add context to what this one aims for exactly.

Thank you a lot again!

yosoyfreeman commented:
I fixed the issues, i think everything is correct now. If that's not the case, tell me and i'll work on it. Thank you!

Mickeon commented on modules/gdscript/editor/script_templates/CharacterBody3D/basic_movement.gd:
```diff
@@ -4,27 +4,38 @@ extends _BASE_
-		velocity.z = move_toward(velocity.z, 0, SPEED)
+		vertical_velocity = up_direction * JUMP_VELOCITY
+
+	# Get the 2D input vector. As good practice, you should replace UI actions with custom gameplay actions.
```
I would normally not say this but given the context it's probably worth to split these in two.
```suggestion
	# Get the 2D input vector.
	# As good practice, you should replace UI actions with custom gameplay actions.
```
Which has one realise, the comment is repeating what the code already says fairly explicitly. So there's no point.
```suggestion
	# As good practice, you should replace UI actions with custom gameplay actions.
```

yosoyfreeman replied:
Thank you for the review! I'll work on those!

About the comments, i added them because i wanted to be a bit more explicit for new users, as i often see them having issues to understand some of this things when they start. However, you are totally right that the code is pretty self explanatory (I tried to separate the steps and use clear variable names for that end). I don't feel particularly strong towards keeping them or removing them, but i do think they can be useful. Maybe we can simplify them so they are less redundant?

In any case, if you think the code is self explanatory enough, i have no issue with fully removing them!

Waiting for your feedback.

Thanks a lot!

Mickeon replied:
I have no suggestions for the moment, but generally speaking code comments are _really_ good when they explain why the code exists, rather than what it does. Unorganized thoughts ahead:

- When getting both the horizontal and vertical velocity, it may be worth explaining why separating the two in the first place. It is unorthodox, and it doesn't help that neither values are `float`, rather they're `Vector3`.
    - Instead of `# Get the vertical velocity (for jump).`, there could be a few details on what project does, instead. May be too much, though.

- There's always exceptions, of course. I think that `# Calculate the intended direction in 3D space.` is a good comment there because the it answers **why** we need the line below, instead of exactly **what** it does step-by-step, if you think about it. The math is also particularly unfriendly, even more so for beginners.

yosoyfreeman replied:
Those are good thoughts! I just pushed a less verbose version of the comments, including your previous recommendations, it looks more clean now. I'm going to revisit the comments and try to see if i can explain exactly why we do each step without adding too much text.

I don't think the math is unfriendly in the context of what it does. I agree that has an inherent complexity but it is the easiest way i could think of to illustrate how to perform correct acceleration/friction on the horizontal axis without it being affected or disturbing the vertical one.

I'm gonna think a bit about it and see if i can make the math more intuitive by explaining the reasoning in the comments instead of being so literal. I think that will help a lot.

Thank you a lot for the help! really appreciate it!

Mickeon replied:
Regarding the math, I didn't mean it should be changed. I wanted to say that summarizing it with a comment is a good thing.

yosoyfreeman replied:
I have this version now, it's a bit more verbose but it may be more clear to read. Before committing, who do you see it against the current one? The issue i have is basically that, to explain properly what is going on it requires a lot more text and context, and that's bad for readability. 

I like this version, but having the cleaner one and a proper documentation page for the templates may be a good thing. I could work on that if needed.

Thank you!

```
func _physics_process(delta: float) -> void:
	# Handle gravity.
	if not is_on_floor():
		velocity += get_gravity() * delta

	# Isoltate the vertical and horizontal components of the velocity so we can prevent
	# any vertical velocity from affecting our horizontal acceleration/friction.
	var vertical_velocity: Vector3 = velocity.project(up_direction)
	var horizontal_velocity: Vector3 = velocity - vertical_velocity

	# Handle jump.
	if Input.is_action_just_pressed("ui_accept") and is_on_floor():
		vertical_velocity = up_direction * JUMP_VELOCITY

	# As good practice, you should replace UI actions with custom gameplay actions.
	var input_vector := Input.get_vector("ui_left", "ui_right", "ui_up", "ui_down", 0.15)

	# Calculate the intended direction in 3D space.
	var input_direction: Vector3 = transform.basis.orthonormalized() * Vector3(input_vector.x, 0, input_vector.y)

	# input_direction is usually a unit vector, but it's magnitude can be less than 1 to allow the
	# use of analog input devices. As such, no special code is needed to handle controller joystic input.
	var target_horizontal_velocity: Vector3 = input_direction * SPEED

	# Move the current horizontal velocity towards the target horizontal velocity.
	horizontal_velocity = horizontal_velocity.move_toward(target_horizontal_velocity, ACCEL * delta)

	# Compose the final velocity.
	velocity = horizontal_velocity + vertical_velocity
```

Mickeon reviewed (commented):
I am not sure if the additional comments help. If anything they make things look more complicated than they are. A lot of this code is good and is readable on its own, it doesn't need the comments repeating themselves.

Gertkeno commented:
Really good fixes to the template. I think it's a great starting point now that it's corrected key issues.

Knockback has been challenging for new developers to make with the preivous templates, I see that as a common forum help post; this template very directly fixes those issues, allowing physics driven forces to affect the character bodies.

A lot of developers would slot animations into the if/else direction: and that remains easy to do, though as obvious. Luckily any old tutorial using the last template should hint at that implementation.

Many do not even notice the old template does not support controllers despite how close it was, just removing `.normalized()` helps.


I am glad nothing else is added, this template should focus on legiblity and being extensible! I could see myself explaining this to a 10 year old easily, skipping over orthonormalized as _rotates the input_ haha.

yosoyfreeman commented:
I just noticed that we are calling jump_velocity to a variable that represents the jump_speed, as speed is just a scalar (what the variable actually is) and velocity implies a vector, which is misleading for the user.

Also, i understand the intention of using constant for the parameters, but things as the speed, the acceleration and the jump speed should be exposed to the editor as i see it.  It would also show the user how to export variables.

What do you think? i personally think that exporting them makes more sense, more having into account that constants in gdscript can be changed anyway.

Hearing your thoughts.

Thank you!

yosoyfreeman commented:
Ok, i pushed the latest changes and i think everything should be right!

This is the final work involved in this PR:

- Fixed the input distortion issue.
- Implemented a correct way to accelerate without square root distortion due each axis using the max acceleration.
- Implemented a variable for acceleration instead of using the max speed.
- Fixed FPS dependency on the acceleration. This is a non trivial issue, but with the current code it's easy to visualize work with. As long as the acceleration is not wrongly manipulated by the user by using some kind of multiplier, we should be OK.

-General refractor of the code.

- Fixes some consistency errors with typed vars.

- Change jump_velocity to jump_speed to reflect that is indeed a scalar, not a vector.

- Exposes speed, accceleration and jump speed to the editor, showing the user how to do it and the correct annotation for both gdscript and c#

I think that's it. I personally like the result a lot and got good feedback, but of course, feel free to tell me if you think it needs further corrections.

sinewavey commented on modules/gdscript/editor/script_templates/CharacterBody3D/basic_movement.gd:
```diff
@@ -2,29 +2,40 @@
-const SPEED = 5.0
-const JUMP_VELOCITY = 4.5
+# The @export annotation allows a variable to be shown and modified from the inspector.
+@export var SPEED: float = 5.0
```
I hate to be a stickler, but these should probably not be capitalized if they're not constants anymore.

yosoyfreeman replied:
Thank you! Not a stickler at all! I left them capitalized because they were being used as constants but you are right, I'll make the correction as soon as possible and be back!

Thank you again!

yosoyfreeman replied:
I just pushed the new changes, fixed the typing to reflect they are now variables as you said and on the way caught an error in the 2D version in which i called JumpSpeed JumPSpeed, so indeed double checking was a good idea haha

I think the typing is now correct. I have few experience with C# but i think it's good now, in any case, if you see anything else that is worth tackling tell me and i'll work on it!

Thank you!
